go/proxyd/backend.go

```diff
@@ -693,3 +697,18 @@ func sleepContext(ctx context.Context, duration time.Duration) {
 	case <-time.After(duration):
 	}
 }
+
+type LimitedHTTPClient struct {
+	http.Client
+	sem         *semaphore.Weighted
+	backendName string
+}
+
+func (c *LimitedHTTPClient) DoLimited(req *http.Request) (*http.Response, error) {
+	if err := c.sem.Acquire(req.Context(), 1); err != nil {
+		tooManyRequestErrorsTotal.WithLabelValues(c.backendName).Inc()
+		return nil, wrapErr(err, "too many requests")
```

Comment:
I think we should return a new `HTTPError` type here that returns a `503` error with a custom JSON-RPC error code and message.

Reply:
Doing so will hide the real reason why we cannot acquire the semaphore. The `Acquire` call blocks until there's an available workload. So an error is returned only when the context is Done, which could be because the user killed the request or the request timed-out. And we already return an appropriate error code for those.

Reply:
Got it.